### PR TITLE
Exercise

### PR DESCRIPTION
**Pull request by z5116890**
- Have cleaned up filename
- endpoints work

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    connection.query('INSERT INTO items (id, todo, completed) VALUES (?,?,?)', [req.body.id, req.body.todo, req.body.completed],(error, 
+        results) => {
+     if (error) return res.json({ error: error });
+     });
+});
+
+const port = process.env.PORT || 4000;
+app.listen(port, () => {
+    console.log(`listening on ${port}`);
+});
```
Comment (paulhartono) [inline server/index.js]: you can delete this file

File: server/src/routes/index.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { Router, Request, Response } from 'express'
+import * as items from './items'
+
+export const setupRoutes = (router: Router): void => {
+
+  router.get('/', (req : Request, res : Response) => {
+    console.log(req.body)
+    items.getTodoItems(req, res)
+  })
```
Comment (paulhartono) [inline server/src/routes/index.ts]: this should be something like: 

```
router.get('/items', items.getTodoItems)
```

File: server/src/routes/index.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  router.delete('/', (req : Request, res : Response) => items.deleteTodoItem(req, res))
+  router.post('/', (req : Request, res : Response) => items.insertTodoItem(req, res))
+  router.patch('/', (req : Request, res : Response) => items.updateTodoItem(req, res))
```
Comment (paulhartono) [inline server/src/routes/index.ts]: same as above.

File: server/src/services/db-operations.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// return all items for item table
+export const getItems = (req : Request, res : Response) => {
+    console.log('hello')
+    console.log(req.body.id)
+    connection.query('select * from items', (error : any, results : any) => {
+        console.log(results)
+        if (error) {
+            return res.json({ error: error })
+        }
+        return res.json(results);
```
Comment (paulhartono) [inline server/src/services/db-operations.ts]: this service function responsibility should only focus on business logic, in this case to interact with database and return the array of items back to the routes. the routes then should be the one returning the json results. 

we can have a call to explain why

File: server/db.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+
+export const connection = mysql.createConnection({
+    host: process.env.DB_HOST,
+    user: process.env.DB_USER,
+    password: process.env.DB_PASS,
+    database: process.env.DB_NAME
+}).on('error', function (err: any) {
+    console.error('mysql failed', err)
+})
```
Comment (paulhartono) [inline server/db.ts]: lets remove this on(error) line and let it break if this throws error - by having this when db connection failed, your app still run

File: server/src/server.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+/* do i need this?
+app.use(router)
+
+app.get('/', (req : Request, res : Response) => items.getTodoItems(req, res))
+app.post('/', (req : Request, res : Response) => items.insertTodoItem(req, res))
+app.delete('/', (req : Request, res : Response) => items.deleteTodoItem(req, res))
+app.patch('/', (req : Request, res : Response) => items.updateTodoItem(req, res))
+
+*/
+const port = process.env.PORT || 4000;
```
Comment (paulhartono) [inline server/src/server.ts]: this `process.env.PORT` will never has value and thus will make port to always be assigned as 4000. Just wanted to see if you know what will be the reason?

File: server/src/services/db-operations.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+     });
+}
+
+//insert item into items table
+export const insertItem = (req : Request, res : Response) => {
+    console.log('hello')
+    connection.query('INSERT INTO items (id, todo, completed) VALUES (?,?,?)', [req.body.id, req.body.todo, req.body.completed],(error : any,
+        results : any) => {
+     if (error) return res.json({ error: error });
```
Comment (paulhartono) [inline server/src/services/db-operations.ts]: similar comment as above - this should be in `routes` not `services`

File: server/src/services/db-operations.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+     });
+
+}
+
+export const deleteItem = function (req : Request, res : Response) {
+    console.log('trying to delete');
+    console.log(req.body.id);
+    connection.query('DELETE FROM items where id = (?)', [req.body.id], function (error : any, results : any) {
+        if (error) return res.json({ error: error });
```
Comment (paulhartono) [inline server/src/services/db-operations.ts]: similar comment as above - this should be in `routes` not `services`

File: server/src/services/db-operations.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        console.log(results)
+    });
+}
+
+export const updateItem = (req : Request, res : Response) => {
+    console.log('updating')
+    console.log(req.body)
+    connection.query('UPDATE items SET completed = NOT completed WHERE id = (?)', [req.body.id],(error : any,
+        results : any) => {
+        if (error) return res.json({ error: error });
```
Comment (paulhartono) [inline server/src/services/db-operations.ts]: similar comment as above - this should be in `routes` not `services`

File: server/src/services/db-operations.ts (hunk 4 above)
Comment (z5116890) [reply server/src/services/db-operations.ts]: Have fixed using async, await and promises but not sure if the way i did it can be simpler

**Comment by z5116890** [comment]
pushed new version with fixes

**Review by paulhartono — CHANGES_REQUESTED** [review]
please try to make the suggested changes

File: server/src/routes/items/delete-todo-item.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import { Request, Response } from 'express'
+import { deleteItem } from '../../services/db-operations'
+import { TodoItem } from '../../model/todo-item';
+
+// get all items from database
+export const deleteTodoItem = (req: Request, res: Response) => {
+  let item: TodoItem = {id: req.body.id, todo: req.body.todo, completed: req.body.completed}
```
Comment (paulhartono) [inline server/src/routes/items/delete-todo-item.ts:7]: just `id` would be required to delete

**Review by paulhartono — APPROVED** [review]

**Comment by paulhartono** [comment]
minor comment for you FYI - you can `squash merge` it

**Comment by z5116890** [comment]
squash merge

**Pull request merged**